File: MergeSortedArray.java
/**
 * LeetCode.88 Merge Sorted Array
 */
class Solution {
    public static void merge(int[] nums1, int m, int[] nums2, int n) {
        if (n == 0) {
            return ;
        }
        if (m == 0) {
            System.arraycopy(nums2,0,nums1,0,n);
            return;
        }
        int l = m + n - 1;

        int j = n - 1;
        int i = m - 1;
        while (j >= 0 && i >= 0) {
            nums1[l--] = (nums1[i] < nums2[j] ? nums2[j--] : nums1[i--]);
        }

        while (i >= 0) {
            nums1[l--] = nums1[i--];
        }

        while (j >= 0) {
            nums1[l--] = nums2[j--];
        }

        return;

    }
}

File: MergeSortedArray.go
func merge(nums1 []int, m int, nums2 []int, n int)  {
	res := make([]int,m + n)

	p1 := 0
	p2 := 0

	cur := 0
	for (p1 < m && p2 < n) {
		if (nums1[p1] > nums2[p2]) {
			res[cur] = nums2[p2]
			p2++
		} else {
			res[cur] = nums1[p1]
			p1++
		}
		cur++
	}

	for (p1 < m) {
		res[cur] = nums1[p1]
		p1++
		cur++
	}

	for (p2 < n) {
		res[cur] = nums2[p2]
		p2++
		cur++
	}

	copy(nums1,res)
	return 
}
